### ISSUE TITLE
[Comparison] ComparisonPerformers buildGetters breaks the filtering

### ISSUE THREAD

**Issue by zuzunker** [vantuz-subhuman/whakamatautau-util#107]
`ComparisonPerformers.buildGetters` adds including filter for public methods. But `GettersDynamicPerformerBuilder` appends a method if **any** of including filters match.

We cannot remove public methods filter from "buildGetters", since:
1. It may break existing code that uses exlusing filters.
2. The usecase is still relevant, we should be able to filter public methods along with other including filters.

The only proper solution is to add functionality to `GettersDynamicPerformerBuilder` to register another kind of filters, that would have `AND` combinator logic with existing filters. Current logic is this:

```
any of (IF1, IF2, IF3, ..., IFn)
AND
none of (EF1, EF2, EF3, ..., EFn)
```

Meaning that method should be accepted by any of including filter and NOT accepted by any of excluding filters. But there should be something like:

```
all of (RF1, RF2, RF3, ..., RFn)
AND
any of (IF1, IF2, IF3, ..., IFn)
AND
none of (EF1, EF2, EF3, ..., EFn)
```

Meaning that method should be accepted by ALL "requirement filters" and then be accepted by any including filter, and also NOT be accepted by any excluding filter. This approach would create 3 distinguished groups of predicates with clear internal combinatory logic. Use case for public methods then would look like this:

```
new GettersDynamicPerformerBuilder<>(type)
    .addRequirement(m -> UberMethods.getVisibility(m) == Visibility.PUBLIC);
```

Meaning that all methods are **required** to be public to be included by any filter. Addition of a second requirement will affect only those methods that have passed the first one. For example:

```
new GettersDynamicPerformerBuilder<>(type)
    .addRequirement(m -> UberMethods.getVisibility(m) == Visibility.PUBLIC)
    .addRequirement(m -> m.getExceptionTypes().length == 0);
```

Meaning that only public methods with no thrown exceptions may be included.
If user will need to include both PUBLIC and PROTECTED methods, for example:

```
new GettersDynamicPerformerBuilder<>(type)
    .addRequirement(m -> UberMethods.getVisibility(m) == Visibility.PUBLIC
        || UberMethods.getVisibility(m) == Visibility.PROTECTED);
```

This new group of methods turns out to be extremely specific, but the requirement is specific itself. Also it seems that there's no need for usability method to add requirement by name regexp, so there may as well be only one method: `adRequirement(Predicate<Method>)`.

**Note:** name `addRequirement` is specified for consistency with existing names `addFilter` and `addExcludingFilter` (which were selected for code readability matters).

**Comment by zuzunker** [comment]
**Note:** there was a discussion to make "requirement filters" inner combinatory logic also as `ANY OF`, but this scheme seems weird and confusing:

```
any of (RF1, RF2, RF3, ..., RFn)
AND
any of (IF1, IF2, IF3, ..., IFn)
AND
none of (EF1, EF2, EF3, ..., EFn)
```

This creates a question: "why there's only two groups of `any of` kind? And is there a way to add more groups like this?"

**Comment by zuzunker** [comment]
**Note:** this fix was backported to `0.0.9` version and released as `0.0.9.1`
https://github.com/Whaka-project/whakamatautau-util/releases/tag/whakamatautau-util-0.0.9.1
